spring-rabbit/src/test/java/org/springframework/amqp/rabbit/annotation/RabbitListenerAnnotationBeanPostProcessorTests.java

```diff
@@ -273,6 +278,42 @@ public void queuesToDeclare() {
 		context.close();
 	}
 
+	@Test
+	public void concurrency() throws InterruptedException, ExecutionException {
+		final int concurrencyLevel = 8;
+		final ExecutorService executorService = Executors.newFixedThreadPool(concurrencyLevel);
+		try {
+			for (int i = 0; i < 1000; ++i) {
```

Comment:
I have modified the test like this:
```
	@Test
	public void concurrency() throws InterruptedException, ExecutionException {
		final ConfigurableApplicationContext context = new AnnotationConfigApplicationContext(Config.class);
		final int concurrencyLevel = 8;
		final ExecutorService executorService = Executors.newFixedThreadPool(concurrencyLevel);
		for (int i = 0; i < 1000; i++) {
			final Callable<?> task = () -> context.getBeanFactory().createBean(BeanForConcurrencyTesting.class);
			final List<? extends Future<?>> futures = executorService
					.invokeAll(Collections.nCopies(concurrencyLevel, task));
			for (Future<?> future : futures) {
				future.get();
			}
		}
		executorService.shutdown();
		context.close();
	}
```
The main point that `new AnnotationConfigApplicationContext(Config.class);` is out of the loop, therefore most performance impact has gone.

Was that really an intention to create a new `applicationContext` on each iteration?
Since the main premise of the issue that concurrently called `createBean()` causes race conditions in the `BPP`.

I understand an aim to catch the issue with as bigger entropy as possible, but there is really no 100 % guaranties that the problem will appear even with the million iterations. 
Typically we test such a race condition with barriers: `CountDownLatch`, `Semaphore` etc...
But I really don't see reason to chase that goal with this fix.

The simplified version of your tests to demonstrate an issue is enough for me.

Just IMHO from my hurt do not spend more time on the fully qualified issue...